drv/auxflash-server/src/main.rs

```diff
@@ -441,17 +441,66 @@ impl idl::InOrderAuxFlashImpl for ServerImpl {
         &mut self,
         _: &RecvMessage,
     ) -> Result<(), RequestError<AuxFlashError>> {
-        (self as &mut ServerImpl)
-            .ensure_redundancy()
-            .map_err(Into::into)
+        ServerImpl::ensure_redundancy(self).map_err(Into::into)
+    }
+
+    fn get_blob_by_tag(
+        &mut self,
+        _: &RecvMessage,
+        tag: [u8; 4],
+    ) -> Result<AuxFlashBlob, RequestError<AuxFlashError>> {
+        self.scan_for_active_slot();
+        let active_slot = self
+            .active_slot
+            .ok_or_else(|| RequestError::from(AuxFlashError::NoActiveSlot))?;
+        let handle = SlotReader {
+            qspi: &self.qspi,
+            base: active_slot * SLOT_SIZE as u32,
+        };
+        let mut outer_reader = TlvcReader::begin(handle)
+            .map_err(|_| AuxFlashError::TlvcReaderBeginFailed)?;
+        while let Ok(Some(outer_chunk)) = outer_reader.next() {
+            if &outer_chunk.header().tag == b"AUXI" {
+                let mut inner_reader = outer_chunk.read_as_chunks();
+                while let Ok(Some(inner_chunk)) = inner_reader.next() {
+                    if inner_chunk.header().tag == tag {
+                        // At this point, the inner reader is positioned *after*
+                        // our target chunk.  We back off by the full length of
+                        // the chunk (including the header), then offset by the
+                        // header size to get to the beginning of the blob data.
+                        let (_, inner_offset, _) = inner_reader.into_inner();
+                        let pos = inner_offset
+                            - inner_chunk.header().total_len_in_bytes() as u64
+                            + core::mem::size_of::<tlvc::ChunkHeader>() as u64;
+                        return Ok(AuxFlashBlob {
+                            slot: active_slot,
+                            start: pos as u32,
+                            end: (pos + inner_chunk.len()) as u32,
```

Comment:
Just checking my understanding:

* `inner_chunk.header().total_len_in_bytes()` is the full length of the chunk (including the header)
* `inner_chunk.len()` is the length of the data in the chunk

Does `total_len_in_bytes()` include more than header+data? If not, could the calculation of `pos` above be `let pos = inner_offset - inner_chunk.len()`?

Reply:
Makes sense :+1: